radix-transactions/src/manifest/static_resource_movements/visitor.rs

```diff
@@ -0,0 +1,724 @@
+        // on whether the resource is fungible or non-fungible.
+        let composite_resource_address =
+            CompositeResourceAddress::from(*state.source_amount.resource_address());
+
+        match (composite_resource_address, state.source_amount) {
+            // Everything on the worktop is being taken so we remove it from the worktop contents.
+            // If the resource was not known to be in the worktop then we create unknown bounds for
+            // it.
+            (
+                CompositeResourceAddress::Fungible(fungible_resource_address),
+                BucketSourceAmount::AllOnWorktop { .. },
+            ) => {
+                self.tracked_buckets.insert(
+                    bucket,
+                    BucketContent::Fungible(
+                        fungible_resource_address,
+                        self.worktop_fungible_contents
+                            .swap_remove(&fungible_resource_address)
+                            .unwrap_or(FungibleBounds {
+                                lower: LowerFungibleBound::NonZero,
+                                upper: UpperFungibleBound::Unbounded,
+                            }),
+                    ),
+                );
+            }
+            (
+                CompositeResourceAddress::NonFungible(non_fungible_resource_address),
+                BucketSourceAmount::AllOnWorktop { .. },
+            ) => {
+                self.tracked_buckets.insert(
+                    bucket,
+                    BucketContent::NonFungible(
+                        non_fungible_resource_address,
+                        self.worktop_non_fungible_contents
+                            .swap_remove(&non_fungible_resource_address)
+                            .unwrap_or(NonFungibleBounds {
+                                amount_bounds: FungibleBounds {
+                                    lower: LowerFungibleBound::NonZero,
+                                    upper: UpperFungibleBound::Unbounded,
+                                },
+                                id_bounds: NonFungibleIdBounds::Unknown,
+                            }),
+                    ),
+                );
+            }
+            // A fungible amount is being taken from the worktop. In the case of fungible resources
+            // the fungible amount will just be deducted from the bounds if bounds are defined. If
+            // the worktop doesn't have knowledge of this resource being here then a bucket with a
+            // guaranteed amount is created,
+            (
+                CompositeResourceAddress::Fungible(fungible_resource_address),
+                BucketSourceAmount::AmountFromWorktop {
+                    amount: bucket_amount,
+                    ..
+                },
+            ) => {
+                // Check if there's an entry for this resource on the worktop. If there is, then we
```

Comment:
(Addressed in https://github.com/radixdlt/radixdlt-scrypto/pull/1931 )

Reply:
> If there's not and there's no uncertainty source, this should be an error.

Good point.

> This would be easier to handle if we just stored the worktop as a FullResourceBounds (as documented in another comment) and we didn't have uncertainty bounds at all. Then we'd do worktop.take_amount(resource, x) and if it was FullResourceBounds::Only and the resource wasn't present, it should throw an error.

I view this as another way to implement the solution to this problem. We can still do this with the model that I have implemented.

Reply:
> I view this as another way to implement the solution to this problem. We can still do this with the model that I have implemented.

Fair :). I think you could implement a lot of the semantic abstractions I've made in your model too. (One of the main purposes of the abstraction is to separate the behind-the-scenes model of the bounds from the business logic in this file - and allow the business logic of the manifest to be simple / verifiable independently from the bounds changes)